File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveElementAccess.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Transforms.InferTypesPlugins.ResolveMethodCalls as resMethCalls
import OneLang.One.Ast.Interfaces as ints
import OneLang.One.Ast.AstTypes as astTypes

class ResolveElementAccess(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveElementAccess")
    
    def can_transform(self, expr):
        is_set = isinstance(expr, exprs.BinaryExpression) and isinstance(expr.left, exprs.ElementAccessExpression) and expr.operator in ["="]
        return isinstance(expr, exprs.ElementAccessExpression) or is_set
    
    def is_map_or_array_type(self, type):
        return astTypes.TypeHelper.is_assignable_to(type, self.main.current_file.literal_types.map) or ArrayHelper.some(lambda x: astTypes.TypeHelper.is_assignable_to(type, x), self.main.current_file.array_types)
    
    def transform(self, expr):
        # TODO: convert ElementAccess to ElementGet and ElementSet expressions
        if isinstance(expr, exprs.BinaryExpression) and isinstance(expr.left, exprs.ElementAccessExpression):
            expr.left.object = self.main.run_plugins_on(expr.left.object)
            if self.is_map_or_array_type(expr.left.object.get_type()):
                #const right = expr.operator === "=" ? expr.right : new BinaryExpression(<Expression>expr.left.clone(), expr.operator === "+=" ? "+" : "-", expr.right);
                return exprs.UnresolvedMethodCallExpression(expr.left.object, "set", [], [expr.left.element_expr, expr.right])
        elif isinstance(expr, exprs.ElementAccessExpression):
            expr.object = self.main.run_plugins_on(expr.object)
            if self.is_map_or_array_type(expr.object.get_type()):
                return exprs.UnresolvedMethodCallExpression(expr.object, "get", [], [expr.element_expr])
            elif isinstance(expr.element_expr, exprs.StringLiteral):
                return exprs.PropertyAccessExpression(expr.object, expr.element_expr.string_value)
        return expr

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveElementAccess.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveElementAccess;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\ElementAccessExpression;
use OneLang\One\Ast\Expressions\UnresolvedMethodCallExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\Expressions\StringLiteral;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Transforms\InferTypesPlugins\ResolveMethodCalls\ResolveMethodCalls;
use OneLang\One\Ast\Interfaces\IType;
use OneLang\One\Ast\AstTypes\TypeHelper;

class ResolveElementAccess extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveElementAccess");
        
    }
    
    function canTransform($expr) {
        $isSet = $expr instanceof BinaryExpression && $expr->left instanceof ElementAccessExpression && in_array($expr->operator, array("="));
        return $expr instanceof ElementAccessExpression || $isSet;
    }
    
    function isMapOrArrayType($type) {
        return TypeHelper::isAssignableTo($type, $this->main->currentFile->literalTypes->map) || \OneLang\Core\ArrayHelper::some($this->main->currentFile->arrayTypes, function ($x) use ($type) { return TypeHelper::isAssignableTo($type, $x); });
    }
    
    function transform($expr) {
        // TODO: convert ElementAccess to ElementGet and ElementSet expressions
        if ($expr instanceof BinaryExpression && $expr->left instanceof ElementAccessExpression) {
            $expr->left->object = $this->main->runPluginsOn($expr->left->object);
            if ($this->isMapOrArrayType($expr->left->object->getType()))
                //const right = expr.operator === "=" ? expr.right : new BinaryExpression(<Expression>expr.left.clone(), expr.operator === "+=" ? "+" : "-", expr.right);
                return new UnresolvedMethodCallExpression($expr->left->object, "set", array(), array($expr->left->elementExpr, $expr->right));
        }
        else if ($expr instanceof ElementAccessExpression) {
            $expr->object = $this->main->runPluginsOn($expr->object);
            if ($this->isMapOrArrayType($expr->object->getType()))
                return new UnresolvedMethodCallExpression($expr->object, "get", array(), array($expr->elementExpr));
            else if ($expr->elementExpr instanceof StringLiteral)
                return new PropertyAccessExpression($expr->object, $expr->elementExpr->stringValue);
        }
        return $expr;
    }
}
